src/common/AriaQueryHandler.ts

```diff
@@ -0,0 +1,122 @@
+    objectId: element._remoteObject.objectId,
+    accessibleName,
+    role,
+  });
+  const filteredNodes: Protocol.Accessibility.AXNode[] = nodes.filter(
+    (node: Protocol.Accessibility.AXNode) => node.role.value !== 'text'
+  );
+  return filteredNodes;
+}
+
+/*
+ * The selectors consist of an accessible name to query for and optionally
+ * further aria attributes on the form `[<attribute>=<value>]`.
+ * Currently, we only support the `name` and `role` attribute.
+ * The following examples showcase how the syntax works wrt. querying:
+ * - 'title[role="heading"]' queries for elements with name 'title' and role 'heading'.
+ * - '[role="img"]' queries for elements with role 'img' and any name.
+ * - 'label' queries for elements with name 'label' and any role.
+ * - '[name=""][role="button"]' queries for elements with no name and role 'button'.
+ */
+type ariaQueryOption = { name?: string; role?: string };
+function parseAriaSelector(selector: string): ariaQueryOption {
+  const normalize = (value: string): string => value.replace(/ +/g, ' ').trim();
+  const knownAttributes = ['name', 'role'];
+  const queryOptions: ariaQueryOption = {};
+  const attributeRegexp = /\[\s*(?<attribute>\w+)\s*=\s*"(?<value>\\.|[^"\\]*)"\s*\]/;
+  const defaultName = selector.replace(
+    attributeRegexp,
+    (_, attribute: string, value: string) => {
+      attribute = attribute.trim();
+      if (!knownAttributes.includes(attribute))
+        throw new Error(
+          'Unkown aria attribute "${groups.attribute}" in selector'
+        );
+      queryOptions[attribute] = normalize(value);
+      return '';
+    }
+  );
+  if (defaultName && !queryOptions.name)
+    queryOptions.name = normalize(defaultName);
+  return queryOptions;
+}
+
+const queryOne = async (
+  element: ElementHandle,
+  selector: string
+): Promise<ElementHandle | null> => {
+  const exeCtx = element.executionContext();
+  const { name, role } = parseAriaSelector(selector);
+  const res = await queryAXTree(exeCtx._client, element, name, role);
+  if (res.length < 1) {
+    return null;
+  }
+  return exeCtx._adoptBackendNodeId(res[0].backendDOMNodeId);
+};
+
+const waitFor = () => {
+  throw new Error('waitForSelector is not supported for aria selectors');
+};
+
+const queryAll = async (
+  element: ElementHandle,
+  selector: string
+): Promise<ElementHandle[]> => {
+  const exeCtx = element.executionContext();
+  const { name, role } = parseAriaSelector(selector);
+  const res = await queryAXTree(exeCtx._client, element, name, role);
+  return Promise.all(
+    res.map((axNode) => exeCtx._adoptBackendNodeId(axNode.backendDOMNodeId))
+  );
+};
+
+const queryAllArray = async (
+  element: ElementHandle,
+  selector: string
+): Promise<JSHandle> => {
+  const elementHandles = await queryAll(element, selector);
+  const exeCtx = element.executionContext();
+  const jsHandle = exeCtx.evaluateHandle(
+    (...elements) => Array.from(elements),
```

Comment:
Since rest parameters always produce arrays, do we need the `Array.from` here?

Reply:
No that looks weird :) Done.